nasa/ow_simulator
# Oceanwater-1140: Re-sync topic publication rate

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-994](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-994) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-1140](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-1140) |
| Github :octocat:  | # |


## Summary of Changes
* The current iteration of OW-994 does not publish prediction data in a timely manner due to the way its main power system loop operates.
  * It sends input data to all asynchronous prognosers at once, but then waits until all prognosers return predictions to continue the loop. This is infeasible when publications are expected at a rate of 0.5Hz, as the return time for 24 prognoser predictions can surpass 20 seconds.
* This PR fully reworks the power system loop so that data is published via rostopics at a rate of 0.5Hz, consistent with the rate of publication in the current single-cell model in _noetic_devel_, even if the prognosers are unable to complete predictions at the same rate.
  * The loop sends input data to the prognosers every cycle, even if they are already generating a prediction, in order to update the state estimate.
  * If a given prognoser has completed its prediction in a cycle before the others, input data is temporarily withheld from that prognoser to prevent triggering another prediction before the others (and thus de-syncing).
  * When all prognosers have completed their predictions, their data is used collectively to update the stored publication values.
  * These stored publication values are published every cycle, so the same values are re-published at a rate of 0.5Hz while waiting for the prognosers to return new predictions.
* Since this change introduces an edge case at the start of the simulation where there is no data to publish for the first several cycles of waiting for prognosers, there are now initial output values published until the prognosers return their first predictions for the simulation. They are as follows:
  * RUL: ``MAX_HORIZON_SECS``. This is the maximum possible published value for RUL, which the prognosers should output for their first predictions at the moment anyways. May need re-evaluation if the max horizon is greatly increased.
  * SoC: ``0.95`` (or 95%). Value was suggested by Chetan.
  * TMP: ``m_initial_temperature``, the same value passed into the prognosers for their first cycle. Currently set to 20C in ``system.cfg``.


## Test
* Check ``ow_power_system/config/system.cfg`` and ensure the ``print_debug`` flag is set to ``true`` while testing so that battery stats are printed to the terminal during runtime. It is ``false`` by default.
  * The subsequent debug flags in the config file can be used to filter out specific parts of the debug statements as long as ``print_debug`` is ``true``. While having all set to ``true`` will create a wall of outputs every 2 seconds, they are important for this test.
* Open up 4 separate terminals
  * In the 1st terminal, build and run the simulator. Environment shouldn't matter, though I use the ``atacama`` world (``roslaunch ow atacama_y1a.launch``).
  * In the other 3 terminals, ``echo`` each one of the published power system topics:
    * ``rostopic echo /battery_remaining_useful_life``
    * ``rostopic echo /battery_state_of_charge``
    * ``rostopic echo /battery_temperature``
  * A very large amount of statements will continually print to the first terminal, assuming the debug flags in ``system.cfg`` were set to ``true``. These statements print on a per-cycle basis, displaying the following in order:
    * The timestamp of the current cycle
    * The input power/voltage/temperature data set to GSAP's asynchronous prognosers
      * Note that if input to a given prognoser is withheld for the cycle, it will not display here. E.g. If there is no line for ``N02``'s inputs here, that means ``N02`` has completed its prediction and is waiting for other prognosers to finish their
    * The values of ``min_rul``, ``min_soc``, and ``max_tmp``. **These are the values that should be published via rostopics for the cycle.**
    * The currently stored output prediction data of RUL/SoC/TMP from GSAP's prognosers, as well as the node's status. It will either be ``PREDICTING``, meaning its prognoser is still generating a prediction, or ``COMPLETE``, meaning it is waiting for the other prognosers to finish their predictions.
* **If everything is running smoothly, the debug statements in the primary terminal will output at a rate of 0.5Hz, and the values seen via rostopics should match the ``min_rul``, ``min_soc``, & ``max_tmp`` values seen in the debug statements.**
* **The loop should follow a pattern of waiting for all nodes to complete their predictions, then updating ``min_rul``/``min_soc``/``max_tmp``, then starting new predictions for each node at the same time. The values reported by ``rostopic echo`` should repeat until a new batch of predictions completes, and they should be receiving values at a rate of 0.5Hz.**
  * **The very first published values (before a set of GSAP predictions completes) should be the same as the initial values provided for RUL/SoC/TMP.**
